src/optimizing/Analyze_rule.ml

```diff
@@ -598,22 +596,31 @@ let regexp_prefilter_of_rule (r : R.rule) =
    * which was triggering a FakeInfoStr exn
    *)
   let key = rule_id in
-  Common.memoized hmemo key (fun () ->
-      try
-        match r.mode with
-        | `Search f
-        | `Extract { formula = f; _ } ->
-            regexp_prefilter_of_formula f
-        | `Taint spec -> regexp_prefilter_of_taint_rule r.R.id spec
-        | `Secrets _ (* TODO *)
-        | `Steps _ ->
-            (* TODO *) None
-      with
-      (* TODO: see tests/rules/tainted-filename.yaml,
-                   tests/rules/kotlin_slow_import.yaml *)
-      | CNF_exploded ->
-          logger#error "CNF size exploded on rule id %s" (rule_id :> string);
-          None
-      | Stack_overflow ->
-          logger#error "Stack overflow on rule id %s" (rule_id :> string);
-          None)
+  let regex_prefilter_fun () =
+    try
+      match r.mode with
+      | `Search f
+      | `Extract { formula = f; _ } ->
+          regexp_prefilter_of_formula f
+      | `Taint spec -> regexp_prefilter_of_taint_rule r.R.id spec
+      | `Secrets _ (* TODO *)
+      | `Steps _ ->
+          (* TODO *) None
+    with
+    (* TODO: see tests/rules/tainted-filename.yaml,
+                 tests/rules/kotlin_slow_import.yaml *)
+    | CNF_exploded ->
+        logger#error "CNF size exploded on rule id %s" (rule_id :> string);
+        None
+    | Stack_overflow ->
+        logger#error "Stack overflow on rule id %s" (rule_id :> string);
+        None
+  in
+  (* Previously, we created hmemo at the toplevel of this file. This caused
+     problems with tests that ended up reusing that table which were very
+     confusing to debug. To prevent that from happening again, the table is
+     now passed to this function. For convenience you can also choose not to
+     memoize. *)
+  match hmemo with
+  | None -> regex_prefilter_fun ()
+  | Some hmemo -> Common.memoized hmemo key regex_prefilter_fun
```

Comment:
I guess the issue is that we use just the Rule_ID.t as the key, and multiple tests can reuse the same rule_ID.t with a different Rule.t, so maybe we should have hashed the Rule.t instead.